chains/ethereum/writer_methods.go

```diff
@@ -244,50 +253,110 @@ func (w *writer) watchThenExecute(m msg.Message, data []byte, dataHash [32]byte,
 		select {
 		case <-w.stop:
 			return
 		default:
-			err := w.conn.LockAndUpdateOpts()
-			if err != nil {
-				w.log.Error("Failed to update tx opts", "err", err)
-				continue
-			}
-			// These store the gas limit and price before a transaction is sent for logging in case of a failure
-			// This declaration is necessary as tx will be nil in the case of an error when sending VoteProposal()
-			// We must also declare variables instead of using w.conn.Opts() directly as the opts are currently locked
-			// here but for all the logging after line 272 the w.conn.Opts() is unlocked and could be changed by another process
-			gasLimit := w.conn.Opts().GasLimit
-			gasPrice := w.conn.Opts().GasPrice
-
-			tx, err := w.bridgeContract.VoteProposal(
-				w.conn.Opts(),
-				uint8(m.Source),
-				uint64(m.DepositNonce),
-				m.ResourceId,
-				dataHash,
-			)
-			w.conn.UnlockOpts()
-
-			if err == nil {
-				w.log.Info("Submitted proposal vote", "tx", tx.Hash(), "src", m.Source, "depositNonce", m.DepositNonce, "gasPrice", tx.GasPrice().String())
-				if w.metrics != nil {
-					w.metrics.VotesSubmitted.Inc()
+			if w.conn.ItxClient() != nil && w.forwarderClient != nil && itxFailures < ItxRetryLimit {
+				forwarderNonce, err := w.forwarderClient.LockAndNextNonce()
+				if err != nil {
+					itxFailures++
+					w.log.Error("Failed to get and lock forwarder nonce for vote proposal", "itxFailures", itxFailures, "err", err)
+					time.Sleep(TxRetryInterval)
+					continue
+				}
+
+				pData, err := packVoteProposalData(
+					uint8(m.Source),
+					uint64(m.DepositNonce),
+					m.ResourceId,
+					dataHash,
+				)
+				if err != nil {
+					itxFailures = ItxRetryLimit
+					w.log.Error("Failed to pack data for vote proposal", "err", err)
+					continue
+				}
+
+				signedData, err := w.forwarderClient.PackAndSignForwarderArg(
+					w.conn.Opts().From,
+					w.cfg.bridgeContract,
+					pData, forwarderNonce,
+					big.NewInt(0),
+					big.NewInt(int64(w.conn.Opts().GasLimit)),
+					uint(m.Destination),
+					*w.conn.Keypair(),
+				)
+				if err != nil {
+					itxFailures = ItxRetryLimit
+					w.log.Error("Failed to sign forwarder data for vote proposal", "err", err)
+					continue
+				}
+
+				// add gas for the forwarder call
+				forwarderGas := uint(w.conn.Opts().GasLimit*64/63 + 100000)
+				signedTx, err := toSignedRelayTx(w.forwarderClient.forwarderAddress.String(), signedData, forwarderGas, uint(uint8(m.Destination)), w.conn.Keypair())
+				if err != nil {
+					itxFailures = ItxRetryLimit
+					w.log.Error("Failed to sign relay tx for vote proposal", "err", err)
+					continue
+				}
+
+				res, err := sendRelayTransaction(w.conn.ItxClient(), w.conn.Opts().Context, signedTx.tx, signedTx.sig)
+				if err != nil {
+					w.forwarderClient.UnlockAndSetNonce(nil)
+					itxFailures++
+					w.log.Error("Failed to send vote proposal to itx", "itxFailures", itxFailures, "err", err)
+					time.Sleep(TxRetryInterval)
+					continue
```

Comment:
`w.proposalIsComplete` is in the `else` block, so removing the `continue` wouldn't mean that it get's called.

`sendRelayTransaction` cannot fail due to on chain state changing - it's only from failures to contact the ITX api. ITX doesnt offer you access to whether or not the transaction will fail. However it sounds like we should check w.proposalIsComplete before sending, rather than after. So I'll add that to the loop.

Reply:
Yes in that case it sounds like a `w.proposalIsComplete` check is not needed. I was referencing adding in a `w.proposalIsComplete` check after a failed `sendRelayTransaction` not trying to access the existing call, but based on how ITX operates with on chain state changes it doesn't matter either way as the check would be unnecessary.